### ISSUE TITLE
Database Update Warning - Is this a Problem?

### ISSUE THREAD

**Issue by IADDIC** [mautic/mautic#6961]
[//]: # ( Invisible comment: 
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Before you create the issue:
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Search for similar report among other reported issues.
Learn how to troubleshoot at https://www.mautic.org/docs/en/tips/troubleshooting.html
Use drag&drop to attach images or other files )

## Bug Description
While migrating and updating Mautic I received the error:

Row size too large. The maximum row size for the used table type, not counting BLOBs, is 65535. This includes storage overhead, check the manual. You have to change some columns to TEXT or BLOBs

| Q   | A
| --- | ---
| Mautic version | 2.14.2
| PHP version | 7.1.24
| Browser | chrone

### Steps to reproduce
1. Using SSH 
2. php app/console doctrine:schema:update --force
 
### Log errors
mautic.NOTICE: Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'ALTER TABLE maujd_leads CHANGE title title VARCHAR(255) DEFAULT NULL, CHANGE firstname firstname VARCHAR(255) DEFAULT NULL, CHANGE lastname lastname VARCHAR(255) DEFAULT NULL, CHANGE company company VARCHAR(255) DEFAULT NULL, CHANGE position position VARCHAR(255) DEFAULT NULL, CHANGE phone phone VARCHAR(255) DEFAULT NULL, CHANGE mobile mobile VARCHAR(255) DEFAULT NULL, CHANGE fax fax VARCHAR(255) DEFAULT NULL, CHANGE

**AND this repeats for every field in my database.**

[//]: # ( Invisible comment:
Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable. )

**Comment by Woeler** [comment]
How many custom lead fields do you have? How wide is your lead table?

This is not a Mautic issue, but an issue with your MYSQL setup.

See here
```
The MySQL maximum row size limit of 65,535 bytes is demonstrated in the following InnoDB and MyISAM examples. The limit is enforced regardless of storage engine, even though the storage engine may be capable of supporting larger rows. 
```

https://dev.mysql.com/doc/refman/8.0/en/column-count-limit.html

**Comment by IADDIC** [comment]
I use Mautic for a business and a charity.  With shopping field and billing field and segmentation and such, we have almost two hundred fields. 

As for a SQL setup problem!  We set Mautic up using Softlicious over 2.5 years ago. All we ever did was update the Database per the Mautic instructions when updating Mautic.  At one time we had an update failure and were told to perform the Mautic schema update via ssh. 

So, some questions, 

1.) Why when we run the schema update are these things not updated to whatever mautic needs?

2.) So, a short question then is how does this get fixed?  

Is there something I can do (some commands) that can be run to universally fix the database?

**Comment by Woeler** [comment]
You have too many custom lead fields, making the lead table too wide. Some migrations in Mautic also add fields to this table to accommodate for new features. Since your table is on the edge of max-width, the migration cannot add the fields it requires for the update.

The only way I see to fix it would be to get rid of some custom fields, then run the migration.

**Comment by IADDIC** [comment]
Well, that certainly changes things.  I do not ever recall seeing or reading about such a limit...we were just "happy campers" adding fields as we thought they were relavent...and most are.  Since this is a SQL limit, I suspect we might be better off running two instances of Mautic, one for each entity.  

Last Question:  Is there a mathematical way of determining how many fields we can have?  I don't mind reducing the number but is there a way to achieve the "magic number" without deleting them one at a time and trying the schema update?  

Thank you for the time you took to investigate this isuue with me.  

Rich

**Comment by Woeler** [comment]
It's not that simple. What it means is that one hypothetical database row with all fields filled to their max capacity may never exceed 65,535 bytes. And at field creation it assumes this hypothetical situation in order to determine if you can add the fields or not.

It really depends on what types of fields you have, and how many of each type. And even those can have an individual configuration.

If you have separate business entities / companies I would highly recommend running one instance per entity, not all in one.

**Comment by IADDIC** [comment]
I think I just dodged a bullet.  I had 5 fields that were recently created and so we decided to delete those first.  After deleting, I re-ran the schema update and although I received thousands of transactions I did not have the blanket error message.  I am taking that to mean at the moment I am in an ok state.

As for running separate instances...I am up for that and it is on the list for next year.  

Thank you very much. I appreciated your help.
